### ISSUE TITLE
connectionFactory.CreateAsync hangs forever

### ISSUE THREAD

**Issue by ajitsinghit2** [Havret/dotnet-activemq-artemis-client#381]
Hi,
When I am trying to make connection from console app, it is working fine but when try the same from webservice, it hangs forever.

Thanks

**Comment by ajitsinghit2** [comment]
Thanks for the quick reply.
I tried connectionFactory.AutomaticRecoveryEnabled = false; but no change.
But to let you know when I put the logic in a thread, it started working. With task working, without it not

System.Threading.Tasks.Task.Run(() => 
{
    activeMQueue.Publish(new MyQueue { JobId = obj.Id});
})


Any though on this.

**Comment by Havret** [comment]
I guess to actually help I would need more comprehensive example. Any chance you could provide MRE (https://en.wikipedia.org/wiki/Minimal_reproducible_example) so I could take a look?

**Comment by ajitsinghit2** [comment]
[Web.zip](https://github.com/Havret/dotnet-activemq-artemis-client/files/9791345/Web.zip)

Here it is

**Comment by Havret** [comment]
Ok, I am back with some results. The problem with your app is that you're blocking on an async call. Which results in a deadlock.

```
[WebMethod(Description = "Send ActiveMQ message")]
public void SendMessage()
{

    ArtemisActiveMQ<JobQueue> activeMQueue = new ArtemisActiveMQ<JobQueue>(
        "localhost",
        61616,
        "mdbprocess",
        "admin",
        "admin");

    activeMQueue.Publish(new JobQueue { JobId = 1 }).Wait(); // that's the problem
}
```

You don't see this issue with your console app, as console apps don't use SynchronizationContext, the same applies to the custom Task that you're spinning. ASP.NET Web Services unfortunately do use it.  

The problem that you're facing is quite nicely described [here](https://blog.stephencleary.com/2012/07/dont-block-on-async-code.html) by Stephen Cleary. 

To make this work you should either rewrite your method to be fully async, more on this [here](https://learn.microsoft.com/en-us/previous-versions/dotnet/articles/aa480516(v=msdn.10)?redirectedfrom=MSDN), or add `ConfigureAwait(false)` to every single async call that's being blocked. 

So, for instance, your `GetConnection` method should look like that:

```
private async Task<IConnection> GetConnection()
{
    var connectionFactory = new ConnectionFactory();
    connectionFactory.AutomaticRecoveryEnabled = false;
    var endpoint = Endpoint.Create(host, port, username, password);
    return await connectionFactory.CreateAsync(endpoint).ConfigureAwait(false);
}

And your `Publish` method like that:

```
public async Task Publish(T message)
{
    string msg = JsonConvert.SerializeObject(message, Formatting.Indented);

    var connection = await GetConnection().ConfigureAwait(false);

    var producer = await connection.CreateProducerAsync(new ProducerConfiguration
    {
        Address = queueName,
        RoutingType = RoutingType.Anycast,
        MessageDurabilityMode = DurabilityMode.Nondurable
    }).ConfigureAwait(false);

    await producer.SendAsync(new Message(msg)).ConfigureAwait(false);

    await connection.DisposeAsync().ConfigureAwait(false);
}
```

While testing your use case I discovered that we also forgot about at least one `ConfigureAwait(false)` call for autorecovery mode of ArtemisNetClient. As a result `ConfigureAwait(false)` trick may not work with ArtemisNetClient (2.10.0) when you set:

```
var connectionFactory = new ConnectionFactory();
connectionFactory.AutomaticRecoveryEnabled = true;
```

I will need to add some analyzer and fix all the async calls to have ConfigureAwait(false) included. It will be released as part of ArtemisNetClient (2.10.1).

Cheers

**Comment by ajitsinghit2** [comment]
Thank you @Havret. This was really helpful.
